### ISSUE TITLE
NLP: LLM: End-to-End Domain-Specifc Chatbot

### ISSUE THREAD

**Issue by sohang3112** [sohang3112/machine-learning-practice#2]
#### Start
- [ ] Choose & download domain data (say medical data)
- [ ] Choose & download local LLM (say huggingface transformers, maybe later Llama360 -- because their training data is full open source, latest others like Llama models are open source but not train data)

LLM inference (NOT via api like Ollama) using:
- [ ] pre-built pipelines of library
- [ ] now do custom pipeline (impl preprocess (tokenize, embeddings, padding), train loops)

Now:
- [ ] **Milestone**: LLM supplemented with basic keyword-based rules (check how good performance can become)

#### Customize on Domain
Fine-Tune / Transfer-Learn (not sure which appropriate, choose) model on domain, make it good enough (also the base llm isn't that good, so this is necessary work, good to learn!). Do using:
- [ ] pre-built pipelines of library
- [ ] custom pipeline (fine tune methods)
- [ ] **Milestone**:  custom LLM (performance significantly better than rule-based!)

#### (optional) Deploy custom LLM on server
- [ ] choose AWS service (EC2 GPU instance should suffice, or maybe Sagemaker -- only if free tier!)
- [ ] deploy using model's lib (NOT with Ollama). host api
- [ ] *ultra optional* very basic web ui (only if existing lib/framework found, otherwise don't bother manually)

#### Revise Theory
- [ ] Should know foundational LLM knowledge: transformers, self-attention, multi-head attention, fine tuning, etc.
